### PR TITLE
updated code with working functionalities.

### PR DESCRIPTION
**Pull request by rishabhjhs**
This code contains the working functionalities of the features asked to add on to the project

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -109,18 +116,43 @@ function resetMoveInterval(){
...
+      console.log(SPEED);
+      speedIntervalId = setInterval(frame, SPEED);
+      clearInterval(generatecarId);                               
+      if(SPEED>=35 && SPEED <=45){
+      generatecarId=setInterval(genetateOponent,SPEED*100);      /*checking if the speed is in this particular range then the setInterval method is called 
+                                                                and generateOponent fxn is passed which will randomly generate cars on random lanes.
+                                                                second argument is passed in reference with the speed so as to maintain the interval of 
+                                                                generated cars accordingly with speed*/                          
+    }
+else if(SPEED>=25 && SPEED<35)
```
Comment (gouravnema) [inline index.js]: Pls name your constants

File: index.js [hunk 2]
```diff
@@ -109,18 +116,43 @@ function resetMoveInterval(){
...
-    console.log(SPEED);
-    speedIntervalId = setInterval(frame, SPEED);
+function resetSpeed(speed)
+    {
+      clearInterval(speedIntervalId);
+      SPEED =speed;
+      console.log(SPEED);
+      speedIntervalId = setInterval(frame, SPEED);
+      clearInterval(generatecarId);                               
+      if(SPEED>=35 && SPEED <=45){
```
Comment (gouravnema) [inline index.js]: pls extract crashlogic in a function

File: index.js [hunk 3]
```diff
@@ -109,18 +116,43 @@ function resetMoveInterval(){
...
+      generatecarId=setInterval(genetateOponent,SPEED*100);      /*checking if the speed is in this particular range then the setInterval method is called
```
Comment (gouravnema) [inline index.js]: pls fix typo though it was in existing code

File: index.js [hunk 4]
```diff
@@ -109,18 +116,43 @@ function resetMoveInterval(){
...
+      generatecarId=setInterval(genetateOponent,SPEED*80);
+    }
+else if(SPEED<5)
+    {
+      generatecarId=setInterval(genetateOponent,300);          //as there is no threshold value of speed that we increase so if speed is below 5  then interval
+                                                             //is called after every 300 milliseconds.
+    }
+else
+    {
+      generatecarId=setInterval(genetateOponent,SPEED*100);   // this condition is written so as to handle the event like if a user starts the game by a down
```
Comment (gouravnema) [inline index.js]: CamelCase

File: index.js [hunk 5]
```diff
@@ -130,11 +162,74 @@ function oponentMoveAnimation(el,lane){
...
+          else if(car_currentPostion==1 && lane=="5vw" && pos==75)
+           {
+                console.log("crash occured");
+                cloneEl.parentElement.removeChild(cloneEl);
+                clearInterval(oponentMoveIntervalId);
+                oponentMoveIntervalId=null;
+                reset_intervals();
+                flag=true; 
+                alert("GAME OVER");
+                window.location.reload();
```
Comment (gouravnema) [inline index.js]: lets not reload. Let's have a reset function which should just reset game state

File: index.js [hunk 6]
```diff
@@ -130,11 +162,74 @@ function oponentMoveAnimation(el,lane){
...
+                
+            }
+
+           else if(car_currentPostion==21 && lane=="25vw" && pos==75)
+            {
+                console.log("crash occured");
```
Comment (gouravnema) [inline index.js]: can we not skip code repeat logic

File: index.js [hunk 7]
```diff
@@ -130,11 +162,74 @@ function oponentMoveAnimation(el,lane){
           cloneEl.style.top = pos + "vh";
           cloneEl.style.left = lane;
           cloneEl.style.zIndex = Date.now() % 1000
-          if(pos >110){
+          var id_of_car = document.getElementById('player');
```
Comment (gouravnema) [inline index.js]: camel case

File: index.js [hunk 8]
```diff
@@ -130,11 +162,74 @@ function oponentMoveAnimation(el,lane){
...
+          var car_currentPostion = parseInt(id_of_car.style.left || 1);    //here the by default value of car_currentposition is changed to 1 insted of 0
+         //console.log("car pos"+car_currentPostion);                     //to handle the crash event in case user just starts the game and do not move the car
```
Comment (gouravnema) [inline index.js]: pls remove commented code

File: index.js [hunk 9]
```diff
@@ -130,11 +162,74 @@ function oponentMoveAnimation(el,lane){
...
-          if(pos >110){
+          var id_of_car = document.getElementById('player');
+          var car_currentPostion = parseInt(id_of_car.style.left || 1);    //here the by default value of car_currentposition is changed to 1 insted of 0
+         //console.log("car pos"+car_currentPostion);                     //to handle the crash event in case user just starts the game and do not move the car     
+         // console.log("pos"+pos);
+
+          //console.log(lane);
+          if(car_currentPostion==11 && lane=="15vw" && pos==75)          //These if else statements are written to handle the crash event if the above 
+          {                                                             //contition satisfy crash will occur and all the interval will close    
+                console.log("crash occured");
```
Comment (gouravnema) [inline index.js]: crash logic is only testing tail to head collision, it is not working on lateral collision.

File: index.js [hunk 10]
```diff
@@ -101,25 +128,49 @@ function carMoveAnimation(els, toPostion){
...
     clearInterval(moveIntervalId);
     moveIntervalId = null;
 }
 
-function resetSpeed(speed){
-    clearInterval(speedIntervalId);
-    SPEED =speed;
-    console.log(SPEED);
-    speedIntervalId = setInterval(frame, SPEED);
+function resetSpeed(s)
```
Comment (gouravnema) [inline index.js:139]: nextSpeed instead of s

File: index.js [hunk 11]
```diff
@@ -130,18 +181,105 @@ function oponentMoveAnimation(el,lane){
...
 
+function resetGame(cloneEl)                        //this function is used to initialise the game
+{
+cloneEl.parentElement.removeChild(cloneEl);
+resetSpeed(50);
+flag=false;
+resetIntervals();
+}
+
+function crashOccurred(carCurrentPostion,lane,pos)
```
Comment (gouravnema) [inline index.js:235]: can we remove entire if else chain. try js objects (hash maps).

File: index.js [hunk 12]
```diff
@@ -130,18 +181,105 @@ function oponentMoveAnimation(el,lane){
...
+    //contition satisfy crash will occur and all the interval will close
+    if(carCurrentPostion==posEleven && lane==middleLane && pos==posSeventySeven)           
+        {                                                                    
+         flag=true;
+          return true; 
+        }
+    else if(carCurrentPostion<=posSeven && lane==leftLane && (pos>=posSeventySeven && pos<=posNintyThree)) 
+        {       
+         flag=true; 
+         return true;
```
Comment (gouravnema) [inline index.js:247]: why setting and using flag(global var) when this function can return required boolean.

**Comment by rishabhjhs** [comment]
Sir as per the logic i have written it actually checks all the conditions
that are needed to have a crash. If we use maps then also we have to get
the values from it and check the condition that is required to satisfy the
crash.

**Comment by rishabhjhs** [comment]
because flag is used to perform an additional function as there are more
the 1 generated opponents and each opponent has its animation id  so to
stop them i have used flag that stops the other opponent apart from the
crashed one.

**Pull request merged**